### ISSUE TITLE
Should `drop_in_place` imply dereferencability?

### ISSUE THREAD

**Issue by JakobDegen** [rust-lang/unsafe-code-guidelines#373]
The following code is currently DB as reported by Miri:

```rust
#[repr(transparent)]
struct LoudDrop(u8);

impl Drop for LoudDrop {
    fn drop(&mut self) {
        println!("{}", self.0);
    }
}

#[repr(C)]
struct PartiallyDrop(LoudDrop, u8);

pub fn main() {
    let mut x = 0;
    let p = std::ptr::addr_of_mut!(x).cast::<PartiallyDrop>();
    
    unsafe {
        core::ptr::drop_in_place(p);   
    }
}
```

Should it be?

**Comment by JakobDegen** [comment]
(I've checked and the LLVM IR we currently emit for this does not have UB)

**Comment by RalfJung** [comment]
If the example had comments, not every reader would have to reverse engineer what the question is about. ;)
The point is that at the `drop_in_place`, `p` points to a `LoudDrop` but at type `PartiallyDrop` is is (partially) dangling beyond the end of the local `x` that it points to.

I think @pcwalton also recently asked to treat the `drop_in_place` argument more like a reference, and I am inclined to agree. Probably the drop_in_place shim should start with a reborrow (and retag) of the argument as a mutable reference?

**Comment by pcwalton** [comment]
By the way, https://doc.rust-lang.org/std/ptr/fn.drop_in_place.html#safety already says that `to_drop` must be *valid* or UB will result, which seems to imply dereferenceable, no? (It doesn't say valid for how many bytes, but I assume it has to mean valid for the entire size of the type, or else that isn't a very useful definition of "valid", as it otherwise would only mean the 1 byte the pointer is pointing at is valid…)

**Comment by JakobDegen** [comment]
That seems like a reasonable reading of the documentation. We'll still want to check how much of the ecosystem actually complies with this though (sorry, didn't get the branch up yesterday)

**Comment by RalfJung** [comment]
Yeah, on the library API front we are probably good, we just haven't made this into language UB yet, at least not in Miri.
